src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs

```diff
@@ -0,0 +1,149 @@
+
+            projectId = _telemetryService.HashValue(project.FullPath);
+        }
+
+        /// <summary>
+        /// Fire a telemetry event when the dependency tree update is complete, 
+        /// conditioned on three heuristics:
+        ///  1. Telemetry is fired with 'Unresolved' label if any target frameworks are
+        ///     yet to process Rules from a Design Time build. This is tracked by 
+        ///     TelemetryState.ObservedDesignTime
+        ///  2. Telemetry is fired with 'Unresolved' label if any relevant Rules from design
+        ///     time builds are yet to report any changes. Relevant Rules are determined
+        ///     based on the 'ICrossTargetRuleHandler's that have processed. This is tracked
+        ///     by TelemetryState.ObservedRuleChanges
+        ///  3. Telemetry fires with 'Resolved' label when 1. and 2. are no longer true
+        ///  4. Telemetry stops firing once all target frameworks observe an Evaluation following
+        ///     the first design time build, or once it has fired once with 'Resolved' label. 
+        ///     This prevents telemetry events caused by project changes and tree updates after initial load has completed.
+        /// </summary>
+        public void ObserveTreeUpdateCompleted()
+        {
+            if (stopTelemetry) return;
+
+            stopTelemetry = isReadyToResolve || IsStopTelemetryInAllTargetFrameworks();
+
+            _telemetryService.PostProperty($"{TelemetryEventName}/{(isReadyToResolve ? ResolvedLabel : UnresolvedLabel)}",
+                ProjectProperty, projectId);
+        }
+
+        public void ObserveUnresolvedRules(ITargetFramework targetFramework, IEnumerable<string> rules)
+        {
+            if (stopTelemetry) return;
+
+            var telemetryState = telemetryStates.GetOrAdd(targetFramework, (key) => new TelemetryState());
+
+            foreach (var rule in rules)
+            {
+                // observe all unresolved rules by default ignoring whether they have actual changes
+                UpdateObservedRuleChanges(telemetryState, rule);
+            }
+        }
+
+        public void ObserveHandlerRulesChanges(
+            ITargetFramework targetFramework,
+            IEnumerable<string> handlerRules, 
+            IImmutableDictionary<string, IProjectChangeDescription> projectChanges)
+        {
+            if (stopTelemetry) return;
+
+            if (telemetryStates.TryGetValue(targetFramework, out var telemetryState))
+            {
+                foreach (var rule in handlerRules)
+                {
+                    var hasChanges = projectChanges.TryGetValue(rule, out var description)
+                        && description.Difference.AnyChanges;
+                    UpdateObservedRuleChanges(telemetryState, rule, hasChanges);
+                }
+            }
+        }
+
+        public void ObserveCompleteHandlers(ITargetFramework targetFramework, RuleHandlerType handlerType)
```

Comment:
>ObserveCompleteHandlers [](start = 20, length = 23)

This whole method's logic as it is, seems to be a bit not stable. The flow is like this: you wait for at least one DTBuild rule to appear to set ObservedDesignTime flag to true, but you are never waiting for ALL DTRules to register with or without changes. It is not race prone here... You could get all evaluations ready, then first DT build comes also ready (with changes for assemblies for example) and you seem to be ready to claim isReadyToResolve = true. But what about other DTBuild rules?

Also what if second TFM's events would start to come after that and set isReadyToResolve back to false?

Reply:
The heuristic here is limited to determining if all the DT rules with a corresponding previously seen Evaluation rule have been observed with changes. This means DT rules without previously seen evaluation may be missed as you observe. That can't be helped since we don't know if any more design time builds are coming. 

So the general heursitic is: observe Evaluation, observe DesignTime, check that all the DT rules expected (from evaluation) have changes. If so, resolve. If not, wait for another DT build. This seems to work most of the time. Worst case scenario is the telemetry event never fires (those can be excluded from the score card), or fires too early.

---
In reply to: [145184725](https://github.com/dotnet/project-system/pull/2879#discussion_r145184725) [](ancestors = 145184725)